packages/entity/src/EntityMutator.ts

```diff
@@ -321,21 +335,170 @@ export class DeleteMutator<
     return await enforceAsyncResult(this.deleteAsync());
   }
 
-  private async deleteInternalAsync(): Promise<Result<void>> {
+  private async deleteInTransactionAsync(): Promise<Result<void>> {
+    return await this.queryContext.runInTransactionIfNotInTransactionAsync((innerQueryContext) =>
+      this.deleteInternalAsync(innerQueryContext)
+    );
+  }
+
+  private async deleteInternalAsync(
+    queryContext: EntityQueryContext,
+    processedEntitiesFromTransitiveDeletions: ReadonlyEntity<any, any, any, any>[] = []
+  ): Promise<Result<void>> {
     const authorizeDeleteResult = await asyncResult(
-      this.privacyPolicy.authorizeDeleteAsync(this.viewerContext, this.queryContext, this.entity)
+      this.privacyPolicy.authorizeDeleteAsync(this.viewerContext, queryContext, this.entity)
     );
     if (!authorizeDeleteResult.ok) {
       return authorizeDeleteResult;
     }
 
-    const id = this.entity.getID();
+    await DeleteMutator.processEntityDeletionForInboundEdgesAsync(
+      this.entity,
+      queryContext,
+      processedEntitiesFromTransitiveDeletions
+    );
 
-    await this.databaseAdapter.deleteAsync(this.queryContext, this.idField, id);
+    await this.databaseAdapter.deleteAsync(
+      queryContext,
+      this.entityConfiguration.idField,
+      this.entity.getID()
+    );
 
-    const entityLoader = this.entityLoaderFactory.forLoad(this.viewerContext, this.queryContext);
+    const entityLoader = this.entityLoaderFactory.forLoad(this.viewerContext, queryContext);
     await entityLoader.invalidateFieldsAsync(this.entity.getAllDatabaseFields());
 
     return result();
   }
+
+  /**
+   * Finds all entities referencing the specified entity and either deletes them, nullifies
+   * their references to the specified entity, or invalidates the cache depending on the
+   * {@link OnDeleteBehavior} of the field referencing the specified entity.
+   *
+   * @remarks
+   * This works by doing reverse fan-out queries:
+   * 1. Load all entity configurations of entity types that reference this type of entity
+   * 2. For each entity configuration, find all fields that contain edges to this type of entity
+   * 3. For each edge field, load all entities with an edge from target entity to this entity via that field
+   * 4. Perform desired OnDeleteBehavior for entities
+   *
+   * @param entity - entity to find all references to
+   */
+  private static async processEntityDeletionForInboundEdgesAsync<
+    TMFields,
+    TMID,
+    TMViewerContext extends ViewerContext,
+    TMEntity extends Entity<TMFields, TMID, TMViewerContext, TMSelectedFields>,
+    TMSelectedFields extends keyof TMFields
+  >(
+    entity: TMEntity,
+    queryContext: EntityQueryContext,
+    processedEntities: ReadonlyEntity<any, any, any, any>[]
```

Comment:
Would suggest a Set over an array here

Reply:
If JS had a way to override object equality I would definitely use a set, but because we need to use a custom equality function for comparing entities (two entities are the same if they have the same type and ID for this purpose) we need to use `.find` anyways and an array makes it clearer that it's not able to do constant time inclusion checks.

Edit: actually I'll just keep a set of the strings used to compare them, good idea.